visitors/assemble.go

```diff
@@ -0,0 +1,359 @@
+			// We have to make sure the state is correct, so we still need to write out
+			// the file header.
+
+			// Set state to valid based on erase polarity
+			// We really should redo the whole header
+			// TODO: Reconstruct header from JSON
+			fh.State = 0x07 ^ uefi.Attributes.ErasePolarity
+			fBuf = f.Buf()
+			fBuf[0x17] = fh.State
+			f.SetBuf(fBuf)
+			return nil
+		}
+
+		// Otherwise, we reconstruct the entire file from the sections and the
+		// file header using data from the JSON. This means that some JSON values
+		// are now respected, including GUID changes. However file lengths and
+		// checksums will be recalculated.
+
+		// Assemble all sections so we know the final file size. We need to do this
+		// to know if we need to use the extended header.
+		fileData := []byte{}
+		dLen := uint64(0)
+		for _, s := range f.Sections {
+			// Align to 4 bytes and extend with 00s
+			// Why is it 00s? I don't know. Everything else has been extended with FFs
+			// but somehow in between sections alignment is done with 0s. What the heck.
+			for count := uefi.Align4(dLen) - dLen; count > 0; count-- {
+				fileData = append(fileData, 0x00)
+			}
+			dLen = uefi.Align4(dLen)
+
+			// Append the section
+			sData := s.Buf()
+			dLen += uint64(len(sData))
+			fileData = append(fileData, sData...)
+		}
+
+		f.SetSize(uefi.FileHeaderMinLength+dLen, true)
+
+		// Set state to valid based on erase polarity
+		fh.State = 0x07 ^ uefi.Attributes.ErasePolarity
+
+		if err = f.ChecksumAndAssemble(fileData); err != nil {
+			return err
+		}
+		return nil
+
+	case *uefi.Section:
+		if len(f.Encapsulated) == 0 {
+			// No children, buffer should already contain data.
+			return nil
+		}
+
+		// Construct the section data
+		secData := []byte{}
+		dLen := uint64(0)
+		for _, es := range f.Encapsulated {
+			// Align to 4 bytes and extend with 00s
+			for count := uefi.Align4(dLen) - dLen; count > 0; count-- {
+				secData = append(secData, 0x00)
+			}
+			dLen = uefi.Align4(dLen)
+
+			esData := es.Value.Buf()
+			dLen += uint64(len(esData))
+			secData = append(secData, esData...)
+		}
+
+		// Special processing for some section types
+		switch f.Header.Type {
+		case uefi.SectionTypeGUIDDefined:
+			ts := f.TypeSpecific.Header.(*uefi.SectionGUIDDefined)
+			if ts.Attributes&uint16(uefi.GUIDEDSectionProcessingRequired) != 0 {
+				var fBuf []byte
+				switch ts.GUID {
+				case uefi.LZMAGUID:
+					fBuf, err = lzma.Encode(secData)
+					f.SetBuf(fBuf)
+					if err != nil {
+						return err
+					}
+				case uefi.LZMAX86GUID:
+					fBuf, err = lzma.EncodeX86(secData)
+					if err != nil {
+						return err
+					}
+				default:
+					return fmt.Errorf("unknown guid defined from section %v, should not have encapsulated sections", f)
+				}
+				f.SetBuf(fBuf)
+			}
+		default:
+			f.SetBuf(secData)
+		}
+
+		// Fix up the header
+		err = f.GenSecHeader()
+
+	case *uefi.FlashDescriptor:
+		if err = f.ParseFlashDescriptor(); err != nil {
```

Comment:
You can just do `err = f.ParseFlashDescriptor()` because you return `err` at the end.